File: quicksort/quick_sort.rb
def quick_sort(list, left = nil, right = nil)
  left ||= 0
  right ||= list.length - 1
  return list if left > right
  i, j = left, right
  pivot = find_pivot(list[left..right])
  p "left #{left}, right #{right}, pivot: #{pivot}, target: #{list[left..right]}}"

  while true
    while(list[i] < pivot) do
      i = i + 1
    end
    while(pivot < list[j] ) do
      j = j - 1
    end
    break if i >= j

    list[i], list[j] = list[j], list[i]
    i = i + 1
    j = j - 1
  end

  list = quick_sort(list, left, i - 1)
  list = quick_sort(list, j + 1, right)
end


def find_pivot(list)
  # p "find:#{list}"
  list[rand(0..list.length-1)]
end

File: quicksort/quicksort.go
package QuickSort

import "fmt"

func QuickSort(list []int, left, right int) {
	fmt.Println("left: right", left, right)
	if left > right {
		return
	}

	i, j := left, right
	pivot := list[j]
	for {
		for list[i] < pivot {
			i++
		}
		for pivot < list[j] {
			j--
		}
		if i >= j {
			break
		}
		if i < j {
			list[i], list[j] = list[j], list[i]
		}
		i++
		j--
	}
	QuickSort(list, left, i-1)
	QuickSort(list, j+1, right)
}

func main() {
	// list := []int{3,4,1,7,9,10,12}
	// return list
	// fmt.Println(QuickSort(list, 0, len(list) -1))
}
